src/Tools/AnalyzerRunner/LooseAssemblyVersionLoader.cs

```diff
@@ -0,0 +1,91 @@
+using System.IO;
+using System.Reflection;
+using System.Runtime.Loader;
+
+namespace AnalyzerRunner
+{
+    internal static class LooseVersionAssemblyLoader
+    {
+        private static readonly Dictionary<string, Assembly> s_pathsToAssemblies = new Dictionary<string, Assembly>(StringComparer.OrdinalIgnoreCase);
+        private static readonly Dictionary<string, Assembly> s_namesToAssemblies = new Dictionary<string, Assembly>();
+
+        private static readonly object s_guard = new object();
+        private static readonly string[] s_extensions = new[] { "ni.dll", "ni.exe", "dll", "exe" };
+
+        /// <summary>
+        /// Register an assembly loader that will load assemblies with higher version than what was requested.
+        /// </summary>
+        public static void Register(string searchPath)
+        {
+            AssemblyLoadContext.Default.Resolving += (AssemblyLoadContext context, AssemblyName assemblyName) =>
+            {
+                lock (s_guard)
+                {
+                    if (s_namesToAssemblies.TryGetValue(assemblyName.FullName, out var assembly))
+                    {
+                        return assembly;
+                    }
+
+                    return TryResolveAssemblyFromPaths(context, assemblyName, searchPath);
+                }
+            };
+        }
+
+        private static Assembly TryResolveAssemblyFromPaths(AssemblyLoadContext context, AssemblyName assemblyName, string searchPath)
+        {
+            foreach (var cultureSubfolder in string.IsNullOrEmpty(assemblyName.CultureName)
+                // If no culture is specified, attempt to load directly from
+                // the known dependency paths.
+                ? new[] { string.Empty }
+                // Search for satellite assemblies in culture subdirectories
+                // of the assembly search directories, but fall back to the
+                // bare search directory if that fails.
+                : new[] { assemblyName.CultureName, string.Empty })
+            {
+                foreach (var extension in s_extensions)
+                {
+                    var candidatePath = Path.Combine(
+                        searchPath, cultureSubfolder, $"{assemblyName.Name}.{extension}");
+
+                    var isAssemblyLoaded = s_pathsToAssemblies.ContainsKey(candidatePath);
+                    if (isAssemblyLoaded || !File.Exists(candidatePath))
+                    {
+                        continue;
+                    }
+
+                    var candidateAssemblyName = AssemblyLoadContext.GetAssemblyName(candidatePath);
+                    if (candidateAssemblyName.Version < assemblyName.Version)
+                    {
+                        continue;
+                    }
+
+                    return LoadAndCache(context, candidatePath);
+                }
+            }
+
+            return null;
+        }
+
+        /// <remarks>
+        /// Assumes we have a lock on _guard.
```

Comment:
Our convention in other parts of the codebase is to suffix the method with _NoLock which indicates "this doesn't take a lock, you had better have done it already".

Reply:
This file is taken verbatim from dotnet/format.

Reply:
Didn't know about that convention. If I am ever back in dotnet/format, I'll work that in. =)

Reply:
Updated naming here to follow Roslyn convention